### ISSUE TITLE
RFC: Support persisted mutations with multipart fetch 

### ISSUE THREAD

**Issue by geekuillaume** [urql-graphql/urql#2961]
## Summary

With the persisted fetch exchange now supporting mutation, it can be used with the multipart fetch exchange to allow uploading files in a persisted fetch operation. Currently, this isn't supported because the persisted fetch exchange will capture the operation and send `null` fields to the server in the persisted operation. The multipart fetch exchange will be skipped for all mutations

## Proposed Solution

I can think of two solutions:
- Integrate the multipart fetch logic inside `persistedFetchExchange` and enable it behind an option flag: This is the simplest to integrate (for me at least) but break the separation of concern between these packages
- Refactor `persistedFetchExchange` to instead of terminating the fetch operation, pass the operation to the forward exchange with the operation hash provided in the operation context: more complex to implement and could potentially create backward incompatible change. I'm also not sure how we can integrate the retry logic of the persisted exchange with an external exchange down the pipeline. 

### Note:

For my usage, this is a blocking problem and I'll need to fix it asap. I will most likely implement the first solution in my fork and can submit a PR with these changes if you think it's a good enough solution. I'm not sure I'll have time to work on the second solution as it needs a lot more in-depth understanding of the codebase and discussion around the implementation.

**Comment by kitten** [comment]
Hiya :wave:

I'm generally not opposed to persisted multipart GraphQL requests in principle. However, there's a concern here I have here that are my biggest gripe with the proposal, and it boils down to maintainability.

The GraphQL over HTTP proposal has become very important to us. In fact, not only have we started moving closer to it, even as it's still changing and only proposed, we've adopted the ["Incremental Delivery"](https://github.com/graphql/graphql-over-http/blob/main/rfcs/IncrementalDelivery.md) proposal and can already support and use the [payload definitions](https://github.com/graphql/graphql-wg/blob/main/rfcs/DeferStream.md#payload-format) for `@defer` and `@stream` built into `urql`.

The question naturally becomes: why are these two proposals built into `@urql/core`? Why are persisted queries and multipart fetch separate?

The obvious first guess would be, neither of these two are official proposals, and hence have a lower chance of being universally adopted.

However, we do have packages for persisted queries and multipart GraphQL requests for file upload because _they are_ almost universally adopted.

That's why the actual and second reason is that they're neither universally and furthermore **they don't have a formal, stable specification.**
This is actually a big deal imho.

Persisted queries aren't such a big deal, but the spec is loosely defined here: https://github.com/apollographql/apollo-link-persisted-queries#apollo-engine

Multipart mutations / GraphQL file upload is defined here: https://github.com/jaydenseric/graphql-multipart-request-spec
(I don't actually recognise this link — I'm not sure if this is the original location... That's part of the problem)

So, now the question is: what is the spec/implementation of a multipart + persisted request?
There is basically no definition for this.

That doesn't mean it's hard to implement, nor is it hard to write a spec for it.
However, I'd say we should be hesitant to creating another loose spec addition without server-side implementors, no chance of adoption, and go around GraphQL over HTTP as a project _and_ then maintain this exchange.

---

To get back to a more tangible point 😅 Sorry for the long explanation...

Sending the `extensions` necessary for persisted queries over the multipart-upload requests isn't rocket science. As long as you can guarantee that your API implements it as you expect, it'll just work.

While this hasn't really happened, we tried anticipating more custom `fetchExchange`s to be created, and that's why `@urql/core/internal` is a public and stable API.

You should be able to use `makeFetchBody`, `makeFetchURL`, `makeFetchOptions`, and `makeFetchSource` and implement this no problem.

(Mostly, the three fetch exchanges we maintain in this repo look more intimidating than they actually are. Once you strip out the options you don't need and the `dispatchDebug` calls, they're quite compact)

So, I'd encourage you to implement this if you need it and I'd happily give guidance as well, if you'd like it. But, I'm hesitant at this point to definitively say that we'd accept a PR adding this as an "officially maintained" exchange package, because I'm not sure we can reasonably maintain something that's supposed to interact with server-side GraphQL implementations, without having a spec or a shared understanding with a kind of "sub-working-group" to actually sustain it as a "proposal".

**Comment by geekuillaume** [comment]
I understand this strategy but I'm not 100% sure I agree it's the most useful for users of URQL. Apollo and relay already seem to take decisions in their implementation and because of this, implement more feature which in turn make them more attractive, grow their user-base, increasing their weight in the GraphQL community and decisions.

Anyway, you're the maintainer of this projects and you should firstly do what you think is the best for the project and manage it however you want to. I know the effort needed to manage something like that and I thank you a lot for this very very good project! :)

I also applaud that you thought about user creating their own exchanges. I've done so to implement a multipart + persisted exchange that works well for my use. It was quite easy thanks to the example of both exchanges and how you architectured the internal code. This exchange is, for simplicity, always using multipart queries and adds the hash of the persisted query in the `operations` field. This works very well with Postgraphile in my case. 

A small improvement could be to more clearly state in the documentation that the exchanges are expected to be customized for each individual usage if needed. It can be scary to jump into creating a new implementation of something internal without knowing the implications in term of complexity. Thanks to your work, the complexity is low and this could be advertised.

I'll close this issue and you can re-open it if a complete spec of multipart + persisted request is made available.

**Comment by kitten** [comment]
Just to provide a quick reply to some of the above,
> Apollo and relay already seem to take decisions in their implementation

Relay actually doesn't. Afaik a network adapter must be written for anything that's out-of-spec or unusual, including file uploads. Apollo has the advantage of basically building against Apollo Server. This means that they have an easy decision to make on any spec deviations or edge cases by simply guaranteeing support against Apollo Server.

---

Re. The issue in general, however, while the spec hasn't been that solid to cover all edge cases (e.g. I was actually looking today at whether it cares what nested `variables` paths are set to when they're bound to be replaced by `File`/`Blob` and it doesn't define this), it _is_ sufficiently supported by many APIs.

What changed my perspective here on this a little is that GraphQL Yoga is actually a very stable target that we wish to build against/for more closely, and GraphQL Yoga and GraphQL Tools' executor actually match our implementation, which is a good signal for this being stable enough to integrate into `@urql/core`

So, basically, the current idea is to build this functionality into `@urql/core` and treat GraphQL Multipart requests as a transport detail, hence allowing spec-extensions to remain exchanges, like `persistedFetchExchange`, that call into our "core" fetch abstraction.

That'll basically allow for persisted+multipart in combination.

https://github.com/urql-graphql/urql/pull/3051
